File: src/action/Fakeread.cc
#include "Fakeread.h"

/* constructor */
Fakeread::Fakeread(int theSensorID, string theNoise, vector<double>& theNoiseParams):Action(FAKEREAD){

	sensorID = theSensorID;
	noise = mapNoise(theNoise);
	for(int i=0; i<(int)theNoiseParams.size(); i++){
		noiseParams.push_back(theNoiseParams[i]);
	}

}


void Fakeread::execute(vector< vector<Noise> >& theNoises, vector< vector< vector<double> > >& theNoisesParams, vector< vector<double> >& theNoisesValues, vector< vector<double> >& theSensorsValues) const{

	//check if null attack
	bool isNullAttack;
	
	switch( (int) noiseParams.size() ){
					   
		case Fakeread::noiseArgc2: // check alpha && beta
			isNullAttack = ( (noiseParams[0] == 1.0) && (noiseParams[1] == 0.0) );
			break;
			
		case Fakeread::noiseArgc3: // check alpha && beta && gamma
			isNullAttack = ( (noiseParams[0] == 1.0) && (noiseParams[1] == 0.0) && (noiseParams[2] == 0.0) );	
			break;
			
		case Fakeread::noiseArgc4: // check alpha && gamma && deltaH && deltaL
			isNullAttack = ( (noiseParams[0] == 1.0) && (noiseParams[1] == 0.0) && (noiseParams[2] == 0.0) && (noiseParams[3] == 0.0) );	
			break;
			
		case Fakeread::noiseArgc5: // check alpha && beta && gamma
			isNullAttack = ( (noiseParams[0] == 1.0) && (noiseParams[1] == 0.0) && (noiseParams[4] == 0.0) );	
			break;

	};
	
	// if null attack remove it if it is present in noises vector @ local filter
	if( isNullAttack ){
		
		for(int i=0; i<(int)theNoises[sensorID].size(); i++){
			
			if( theNoises[sensorID][i] == noise ){
				theNoises[sensorID].erase( theNoises[sensorID].begin() + i );
				theNoisesParams[sensorID].erase( theNoisesParams[sensorID].begin() + i );
				theNoisesValues[sensorID].erase( theNoisesValues[sensorID].begin() + i );
			}
			
		}	
	
	}
	
	// if not null attack, overwrite it if it is present in the noises vector @ local filter, otherwise add it
	else{
		
		bool isAttackPresent = false;

		for( int i=0; i<(int)theNoises[sensorID].size(); i++){
		
			if( theNoises[sensorID][i] == noise ){ // attack founded, overwrite it
			
				isAttackPresent = true;
				
				// overwrite the noises parameters stored in the noisesParameters vector @ local filter
				for(int j=0; j<(int)theNoisesParams[sensorID][i].size(); j++){
					theNoisesParams[sensorID][i][j] = noiseParams[j];				
				}			
				
				// overwrite the noises values stored in the noisesValues @ local filter
				switch(noise){
								
					case FIX: // FIX fixes the last sensed value at occurrence time of the attack
						theNoisesValues[sensorID][i] = theSensorsValues[sensorID][0]; 
						break;
					
					case MIN: // MIN fixes the minimum sensed value at occurrence time of the attack (i.e. inf)
						theNoisesValues[sensorID][i] = theSensorsValues[sensorID][4]; 
						break;
					
					case MAX: // MAX fixes the maximum sensed value at occurrence time of the attack (i.e. sup)
						theNoisesValues[sensorID][i] = theSensorsValues[sensorID][5]; 
						break;
					
					case AVG: // AVG fixes the average of the sensed value at occurrence time of the attack (i.e. med)
						theNoisesValues[sensorID][i] = theSensorsValues[sensorID][6]; 
						break;
									
					default:
						theNoisesValues[sensorID][i] = 0;
						break;
						
				};
				
			} // end if
		
		} // end for
		
		// add the attack
		if(!isAttackPresent){
		
			// add the noise in the noises vector @ local filter
			theNoises[sensorID].push_back(noise);
			
			// add the noise params in the noisesParams vector @ local filter
			theNoisesParams[sensorID].push_back( vector<double>() );
			for( int i=0; i<(int)noiseParams.size(); i++){
				theNoisesParams[sensorID].back().push_back( noiseParams[i] );
			}
			
			// add the noise values in the noisesValues vector @ local filter
			switch(noise){
			
				case FIX: 
					theNoisesValues[sensorID].push_back( theSensorsValues[sensorID][0]);
					break;
				
				case MIN:
					theNoisesValues[sensorID].push_back( theSensorsValues[sensorID][4]);
					break;
					
				case MAX:
					theNoisesValues[sensorID].push_back( theSensorsValues[sensorID][5]);
					break;
				
				case AVG:
					theNoisesValues[sensorID].push_back( theSensorsValues[sensorID][6]);
					break;
			
				default: 
					theNoisesValues[sensorID].push_back(0);
					break;
			};
		
		}
		
	} // end else
	

}


/* maps the noise function name (string) in NOISE type */
Noise Fakeread::mapNoise(string str){

	if(str == "con"){
		return CON;
	}

	if(str == "sen"){
		return SEN;
	}

	if(str == "sat"){
		return SAT;
	}

	if(str == "fix"){
		return FIX;
	}
	
	if(str == "min"){
		return MIN;
	}
	
	if(str == "max"){
		return MAX;
	}

	if(str == "avg"){
		return AVG;
	}

	if(str == "rnd"){
		return RND;
	}

	if(str == "inf"){
		return INF;
	}

	if(str == "sup"){
		return SUP;
	}

	if(str == "med"){
		return MED;
	}

	if(str == "sgn"){
		return SGN;
	}

	if(str == "sym"){
		return SYM;
	}

	if(str == "shp"){
		return SHP;
	}

	if(str == "lin"){
		return LIN;
	}

	if(str == "sin"){
		return SIN;
	}

	if(str == "sqr"){
		return SQR;
	}

	if(str == "saw"){
		return SAW;
	}

	if(str == "tri"){
		return TRI;
	}
	
	cerr<<"Fakeread::mapNoise -> noise function not recognized\n";

}

File: interpreter/interpreter/primitives/fakeread.py
"""

Fakeread class models the 'fakeread' primitive. It tampers the readings 
of the sensors.

Fakeread is an overloaded primitive
The sintaxes of the primitive are:
 1) fakeread(nodeId, time, sensorId, noise, alpha, beta)
 2) fakeread(nodeId, time, sensorId, noise, alpha, beta, gamma)
 3) fakeread(nodeId, time, sensorId, noise, alpha, deltaH, deltaL, gamma)
 4) fakeread(nodeId, time, sensorId, noise, alpha, beta, tau, phi, gamma)
 + node Id is the Id of the target node
 + time is the occurrence time
 + sensorId is the target sensor of the target node
 + noise is the additive noise function to apply to the sensor reading
 + alpha, beta, gamma, deltaH, deltaL, tau, phi are parameters related to the noise functions
 
NB: nodeId and occurrence time are not stored in this object.

Author:
 + Francesco Racciatti	<racciatti.francesco@gmail.com>

Maintainer:
 + Francesco Racciatti <racciatti.francesco@gmail.com>

"""


from utilities import wrong_argc


class Fakeread:
    """
    fakeread(nodeId, time, sensorId, noise, alpha, beta)
    fakeread(nodeId, time, sensorId, noise, alpha, beta, gamma)
    fakeread(nodeId, time, sensorId, noise, alpha, deltaH, deltaL, gamma)
    fakeread(nodeId, time, sensorId, noise, alpha, beta, tau, phi, gamma)
    nodeId and time are not stored in this object
    """
    
    argv = []
    name = "Fakeread"
    
    argc4 = 4 # sensorID, noise, alpha, beta
    argc5 = 5 # sensorID, noise, alpha, beta, gamma
    argc6 = 6 # sensorID, noise, aplha, deltaH, deltaL, gamma
    argc7 = 7 # sensorID, noise, alpha, beta, tau, phi, gamma
    argc4noises = ['con']
    argc5noises = ['sen', 'sat', 'fix' ,'min', 'max', 'avg', 'inf', 'sup', 'med', 'sgn', 'rnd', 'lin']
    argc6noises = ['shp', 'sym']
    argc7noises = ['sin', 'sqr', 'saw', 'tri']
    
    def __init__(self, args):
        self.argv = args.split(":")
    
    def __str__(self):
        
        # check the number of arguments
        if (len(self.argv) == self.argc4) and (self.argv[1] in self.argc4noises):
            sensorID = self.argv[0]
            noise = self.argv[1]
            alpha = self.argv[2]
            beta = self.argv[3]
            
            params = sensorID + ":" + noise + ":" + alpha + ":" + beta
            xml = "\n\t\t\t<action>"
            xml += "\n\t\t\t\t<name>Fakeread</name>"
            xml += "\n\t\t\t\t<parameters>" + params + "</parameters>"
            xml += "\n\t\t\t</action>"
            
            return xml
        
        elif (len(self.argv) == self.argc5) and (self.argv[1] in self.argc5noises):
            sensorID = self.argv[0]
            noise = self.argv[1]
            alpha = self.argv[2]
            beta = self.argv[3]
            gamma = self.argv[4]
            
            params = sensorID + ":" + noise + ":" + alpha + ":" + beta + ":" + gamma
            xml = "\n\t\t\t<action>"
            xml += "\n\t\t\t\t<name>Fakeread</name>"
            xml += "\n\t\t\t\t<parameters>" + params + "</parameters>"
            xml += "\n\t\t\t</action>"
            
            return xml
        
        elif (len(self.argv) == self.argc6) and (self.argv[1] in self.argc6noises):
            sensorID = self.argv[0]
            noise = self.argv[1]
            alpha = self.argv[2]
            gamma = self.argv[3]
            deltaH = self.argv[4]
            deltaL = self.argv[5]
            
            params = sensorID + ":" + noise + ":" + alpha + ":" + gamma + ":" + deltaH + ":" + deltaL
            xml = "\n\t\t\t<action>"
            xml += "\n\t\t\t\t<name>Fakeread</name>"
            xml += "\n\t\t\t\t<parameters>" + params + "</parameters>"
            xml += "\n\t\t\t</action>"
            
            return xml
        
        elif (len(self.argv) == self.argc7) and (self.argv[1] in self.argc7noises):		
            sensorID = self.argv[0]
            noise = self.argv[1]
            alpha = self.argv[2]
            beta = self.argv[3]
            tau = self.argv[4]
            phi = self.argv[5]
            gamma = self.argv[6]
            
            params = sensorID + ":" + noise + ":" + alpha + ":" + beta + ":" + tau + ":" + phi + ":" + gamma
            xml = "\n\t\t\t<action>"
            xml += "\n\t\t\t\t<name>Fakeread</name>"
            xml += "\n\t\t\t\t<parameters>" + params + "</parameters>"
            xml += "\n\t\t\t</action>"
            
            return xml
        
        
        else:
            wrong_argc(self.name, len(self.argv))
